Fix build on kernel 6.17

x5444 opened the pull request:
Hi @leifliddy, hope you're doing well these days!
This PR solves #43 but it certainly breaks compatibility with older kernels and I don't currently see a good way of making it work with both given how much was changed in the hda folder and file structure.

Basically, I rebased the changes you made to patch_cirrus.c onto the new cs420x.c.
Beyond that I had to change...
- ...the `codec->patch_ops` line (cs420x.c L735) due to this change: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/include/sound/hda_codec.h?h=v6.17.3&id=6bf917e9aacc360e76b74249209a4290f2e8404b
- ...the `free` callback to `remove` due to this change: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/sound/hda/codecs/generic.c?h=v6.17.3&id=cabaf5908e586156d1d59f3fb8c51d2b169bc636

I didn't do all that much research into these changes so I hope I didn't miss anything about the new callbacks/structs.

Can you think of a way to do this without breaking the driver on previous kernels?

Commit:
switch to the new upstream structure of hda source files

Commit:
remove patches made obsolete by new dependency on kernel >=6.17

ejcupcake commented:
Just used your fork on Kubuntu.
<img width="1423" height="725" alt="image" src="https://github.com/user-attachments/assets/d51f5ebe-39e5-4765-b989-93802dd9e1aa" />

The main thing I noticed is that the `-i` flag does not work. It only works when ran without any flags. Works great otherwise.

Commit:
re-add patch_cirrus.c and restore compatibility with older kernels

x5444 commented:
I updated my branch to restore compatibility with older kernels. I had to duplicate some files, if someone can think of a better way, please let me know. I'd greatly appreciate someone trying to run this on an older kernel (<6.17) just to confirm.

I also added the cut operation necessary to fix the driver on Void Linux, this should hopefully fix #40. Currently it would still break for kernels 6.12 up to including 6.12.4 (as has been the case even before this PR).

@ejcupcake thank you for reporting the dkms issue. I haven't looked into it yet but it seems that some more kernel version switching is necessary to get this to run.

leifliddy commented:
@x5444 Nice work, thanks!
